### PR TITLE
Fix problems with `kw vars`

### LINKED ISSUE #142: `kw v` doesn't show all configurations
Possible solution: just iterate through `$configurations` and print its entries.
Downside: will also possibly print the configurations not used by kw (if user defines an unknown setting).

### PR DESCRIPTION
**Pull request by joaoseckler**
This PR includes two commits regarding `kw vars`:
 - `574037c` The command was not showing every configuration option. Now it shows all configuration options which are defined. Closes #142.
 - `580bf26` The abbreviation `kw v` was removed in the code when merging #169, but the documentation was never updated. This commit does that.

@matheustavares suggested as a possible solution for #142 to iterate over configurations and print it's entries. Besides the downside he described himself for this approach (printing mistakenly user-defined options) I would add that it doesn't allow a meaningful description of the options. The downside of hard-coding options in `src/kw_config_loader.sh` (our approach here) is that every new option added to `kworkflow_template.config` needs to be implemented in `src/kw_config_loader.sh` too. For developers to remember to do so, this PR also introduces a test to check whether the options in `etc/kworkflow_template.config` match the options defined in `src/kw_config_loader.sh`.

### CODE REVIEW

File: src/kw_config_loader.sh [hunk 1]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    test_mode='1'
+  fi
+
+  local -A descriptions=(
+    [arch]='Target arch'
+    [mount_point]='Mount point'
+    [virtualizer]='Virtualization tool'
+    [qemu_hw_options]='QEMU options'
+    [qemu_net_options]='QEMU Net options'
+    [ssh_ip]='Default ssh ip'
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:41]: how about "SSH address".

File: src/kw_config_loader.sh [hunk 2]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    [ssh_port]='Default ssh port'
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:42]: SSH Port

File: src/kw_config_loader.sh [hunk 3]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    [kernel_img_name]='Kernel image name'
+    [cross_compile]='Cross-compile name'
+    [menu_config]='Default kernel menu config'
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:45]: Kernel menu config

File: src/kw_config_loader.sh [hunk 4]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    [qemu_path_image]='Default virtualization tool'
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:46]: That's the QEMU image path, not a "virtualization tool"

File: src/kw_config_loader.sh [hunk 5]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    [alert]='Default alert options'
+    [sound_alert_command]='Command for sound completion alert'
+    [visual_alert_command]='Command for visual notification of completion'
+    [doc_type]='Command to generate kernel-doc'
+    [default_deploy_target]='Deploy default target'
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:51]: Target deploy

File: src/kw_config_loader.sh [hunk 6]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
   echo -e "\tLocal config file: $has_local_config_path"
-  echo -e "\tTarget arch: ${configurations[arch]}"
-  echo -e "\tMount point: ${configurations[mount_point]}"
-  echo -e "\tVirtualization tool: ${configurations[virtualizer]}"
-  echo -e "\tQEMU options: ${configurations[qemu_hw_options]}"
-  echo -e "\tQEMU Net options: ${configurations[qemu_net_options]}"
-  echo -e "\tVdisk: ${configurations[qemu_path_image]}"
+
+  for option in "${!descriptions[@]}"; do
+    if [[ -v configurations[$option] || "$test_mode" == '1' ]]; then
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:62]: Prefer "${configurations[$option]}"

File: src/kw_config_loader.sh [hunk 7]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    [visual_alert_command]='Command for visual notification of completion'
+    [doc_type]='Command to generate kernel-doc'
+    [default_deploy_target]='Deploy default target'
+    [reboot_after_deploy]='Reboot after deploy'
+    [disable_statistics_data_track]='Disable tracking of statistical data'
+    [gui_on]='Command to activate GUI'
+    [gui_off]='Command to deactivate GUI'
+  )
+
   say "Variables:"
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:58]: Maybe we can say something like "kw configuration values"

File: src/kw_config_loader.sh [hunk 8]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
-  echo -e "\tTarget arch: ${configurations[arch]}"
-  echo -e "\tMount point: ${configurations[mount_point]}"
-  echo -e "\tVirtualization tool: ${configurations[virtualizer]}"
-  echo -e "\tQEMU options: ${configurations[qemu_hw_options]}"
-  echo -e "\tQEMU Net options: ${configurations[qemu_net_options]}"
-  echo -e "\tVdisk: ${configurations[qemu_path_image]}"
+
+  for option in "${!descriptions[@]}"; do
+    if [[ -v configurations[$option] || "$test_mode" == '1' ]]; then
+      echo -e "\t${descriptions[$option]} ("$option"): ${configurations[$option]}"
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:63]: IMO, replace `\t` for two spaces will look better.

File: tests/kw_config_loader_test.sh [hunk 9]
```diff
@@ -12,6 +12,7 @@ function suite
   suite_addTest "parse_configuration_output_Test"
   suite_addTest "parse_configuration_standard_config_Test"
   suite_addTest "parse_configuration_files_loading_order_Test"
+  suite_addTest "test_show_variables"
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:15]: I thought we gonna drop `suite_addTest`...

File: tests/kw_config_loader_test.sh [hunk 10]
```diff
@@ -138,4 +139,34 @@ $PWD/$CONFIG_FILENAME"
   assertTrue "Wrong config file reading order.\n$expected_vs_got" $?
 }
 
+function test_show_variables() {
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:142]: `{` should be in a new line

File: tests/kw_config_loader_test.sh [hunk 11]
```diff
@@ -138,4 +139,34 @@ $PWD/$CONFIG_FILENAME"
...
+  local showed_options
+
+  # get all assigned options, including commented ones
+  # remove #'s and ='s to get option names
+  local -A possible_options
+  for o in $(cat "etc/kworkflow_template.config" \
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:148]: For the test, use files from `tests/samples/`. Feel free to update it if you need.

File: tests/kw_config_loader_test.sh [hunk 12]
```diff
@@ -138,4 +139,34 @@ $PWD/$CONFIG_FILENAME"
...
+    | grep -oE '^(#?\w+=?)' \
+    | sed  -E 's/[#=]//g'); do
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:150]: I think this code will be easy to read if you put it before this `for` loop.

File: tests/kw_config_loader_test.sh [hunk 13]
```diff
@@ -138,4 +139,34 @@ $PWD/$CONFIG_FILENAME"
...
+    possible_options[$o]='1'
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:151]: Double-quote around `$o`. Also, `o` is a bad name...

File: tests/kw_config_loader_test.sh [hunk 14]
```diff
@@ -138,4 +139,34 @@ $PWD/$CONFIG_FILENAME"
...
+  done
+
+  local -A shown_options
+  for o in $(show_variables 'TEST_MODE' | tail +3 | sed 's/.*(\(\S*\)).*/\1/'); do
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:155]: Move the code after `in` before this `for`.

File: src/kw_config_loader.sh [hunk 15]
```diff
@@ -27,14 +28,41 @@ function show_variables()
     has_local_config_path="No"
   fi
 
+  if [[ "$1" == 'TEST_MODE' ]]; then
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:31]: I did not understand the value of TEST_MODE in this function.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @joaoseckler ,

In general, this PR looks really good, I left a couple of comments in the code, but I also have the following comments:

* When you print the information, you just display it in a random order, making it hard to understand the output. Could you group the information? For example, show ssh info, followed by build info, qemu...
* Can we have a test that checks the displayed option? I was thinking about a test that will call show_variables, grep the value after `:` and compare with an expected sequence.

File: src/kw_config_loader.sh [hunk 16]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+    test_mode='1'
```
[the remark this answers is not in the document]
Comment (joaoseckler) [reply src/kw_config_loader.sh:32]: I was following the convention of single-quoting strings without expansions... That doesn't apply here?

File: src/kw_config_loader.sh [hunk 17]
```diff
@@ -27,14 +28,41 @@ function show_variables()
...
+  fi
+
+  local -A descriptions=(
+    [arch]='Target arch'
+    [mount_point]='Mount point'
+    [virtualizer]='Virtualization tool'
```
[the remark this answers is not in the document]
Comment (joaoseckler) [reply src/kw_config_loader.sh:38]: Hm, `virtualizer` doesn't seem to be the notification tool, check this line from `kworkflow_template.config` out:

`virtualizer=qemu-system-x86_64`

Maybe `VM tool` instead?

File: src/kw_config_loader.sh (hunk 4 above)
Comment (joaoseckler) [reply src/kw_config_loader.sh:46]: Oops, that was a copy-paste I forgot to alter later...

File: src/kw_config_loader.sh (hunk 6 above)
Comment (joaoseckler) [reply src/kw_config_loader.sh:62]: There is a semantic difference here:

`[[ -v "${configurations[$option]}" ]]` checks whether the variable whose name is defined by the expansion of `"${configurations[$option]}"` is set;

`[[ -v configurations["$option"] ]]` check wheter the variable itself if set.

But the quotes around `$option` were indeed missing

File: src/kw_config_loader.sh (hunk 8 above)
Comment (joaoseckler) [reply src/kw_config_loader.sh:63]: agree!

File: tests/kw_config_loader_test.sh (hunk 9 above)
Comment (joaoseckler) [reply tests/kw_config_loader_test.sh:15]: I think so too! But only after #312 is merged... The problem is that if a `suite` function is defined, shunit2 will not automatically run functions which start with `test`. I thought it was better to leave the refactoring of the entire test code to #312.

File: tests/kw_config_loader_test.sh (hunk 11 above)
Comment (joaoseckler) [reply tests/kw_config_loader_test.sh:148]: Makes sense... but it seems like this test is made for developers not to forget to update `kw vars` when they add a new option to `kworkflow_template.config`. Making a copy of it in `tests/samples` would defeat this purpose... What do you think?

File: src/kw_config_loader.sh (hunk 15 above)
Comment (joaoseckler) [reply src/kw_config_loader.sh:31]: Test mode assures all possible configurations are printed, regardless if they are defined or not.

File: src/kw_config_loader.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply src/kw_config_loader.sh:32]: I think we don't need quotes around numbers, we can use the plain number. But you're are right, this is not clear; we probably need to update our codestyle for that.

File: src/kw_config_loader.sh (hunk 17 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply src/kw_config_loader.sh:38]: You're right, I missed that. VM tool sounds good.

File: tests/kw_config_loader_test.sh (hunk 11 above)
Comment (rodrigosiqueira) [reply tests/kw_config_loader_test.sh:148]: That's a good point, you're right.

File: src/kw_config_loader.sh (hunk 15 above)
Comment (rodrigosiqueira) [reply src/kw_config_loader.sh:31]: Usually, I use the test mode for validating command sequences or other things that might be hard to test. Here, I'm struggling to see the value of having this TEST_MODE. For example, take a look at some of the tests associated with mk.

File: tests/kw_config_loader_test.sh (hunk 11 above)
Comment (joaoseckler) [reply tests/kw_config_loader_test.sh:148]: For example, right now there is an old version of `kworkflow_template.config` in the `tests/samples` folder.

File: src/kw_config_loader.sh (hunk 15 above)
Comment (joaoseckler) [reply src/kw_config_loader.sh:31]: Our first approach for this test was to define a custom `configurations` array, including every possible valid configuration option for kw. This way we can test if all the valid options are being printed by `show_variables`, by grepping it's output. Now we check that the set of configuration options that `show_variables` is capable of printing is not smaller than the set of valid configuration options.

Consider the case where a configuration option is removed from kw, and is no longer valid. Consider that the developer who removed that option forgot to remove it from `show_variables`. Now if a user defines this obsolete option manually in his `kworkflow.config`, `kw vars` will still print it as a valid option! We want to avoid that and we want our test to cover this case. (keep in mind that if an option is not defined, it will not be printed by `show_variables`). But how can we check if some undefined option could potentially be printed by `show_variables`? That is, how can we guarantee that the set of configuration option that `show_variables` is capable of printing is not bigger, i.e., does not contain more elements than the set of valid configuration options for kw? We would have to test every possible option, but that is not feasible. This is the purpose of TEST_MODE: it alters the default behavior of not printing options if they are undefined: in TEST_MODE, `show_variables` will print every configuration option that it is capable of printing. That way we can test this edge case.

**Comment by joaoseckler** [comment]
The latest update to this PR includes:
 - Grouping of configuration options by type (ssh, build, deploy, etc)
 - Implementation of a "correctness" test, in addition to the previous "completeness" test
 - formatting and clearer code

File: tests/kw_config_loader_test.sh [hunk 18]
```diff
@@ -141,4 +143,84 @@ $PWD/$CONFIG_FILENAME"
...
+  # remove #'s and ='s to get option names
+  output="$(cat 'etc/kworkflow_template.config')"
+  output="$(echo "$output" | grep -oE '^(#?\w+=?)' | sed -E 's/[#=]//g')"
+
+  for option in $output; do
+    possible_options["$option"]='1'
+  done
+
+  output="$(show_variables 'TEST_MODE' | grep -E '^    ')"
+  output="$(echo "$output" | sed 's/.*(\(\S*\)).*/\1/')"
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:162]: I think we can ignore this shellcheck warning. However, it is worth checking if you can simplify...

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
After reviewing this patch, I noticed that your approach would require a code update for every new option in the config file. In this sense, what do you think about this approach:

Let's update kworkflow.config in a way that we enforce a basic syntax. Something like this:

# Section: Kernel build options
# Summary: Target arch
arch=x86_64
# Summary: Kernel image name
kernel_img_name=bzImage
...

You will also need to reorganize all kworkflow.config section a logical way. After that, instead of creating arrays, you can parser the file.

What do you think?

**Comment by joaoseckler** [comment]
> After reviewing this patch, I noticed that your approach would require a code update for every new option in the config file. In this sense, what do you think about this approach:
> 
> Let's update kworkflow.config in a way that we enforce a basic syntax. Something like this:
> 
> ```
> # Section: Kernel build options
> # Summary: Target arch
> arch=x86_64
> # Summary: Kernel image name
> kernel_img_name=bzImage
> ...
> ```
> 
> You will also need to reorganize all kworkflow.config section a logical way. After that, instead of creating arrays, you can parser the file.
> 
> What do you think?

I think it is a good idea, but I have a few remarks:

- Should we impose another syntax to a configuration file which could be as as simple as possible? 
- Wouldn't it make `kworkflow.config` less user friendly?
- Your suggestion would imply a bigger feature and demand a bit of time, which I think is not our priority right now;
- The test is there to help developers remember to update the function whenever they update `kworkflow_template.config` (which shouldn't occur too often, as well)

What do you think?

**Comment by rodrigosiqueira** [comment]
> > After reviewing this patch, I noticed that your approach would require a code update for every new option in the config file. In this sense, what do you think about this approach:
> > Let's update kworkflow.config in a way that we enforce a basic syntax. Something like this:
> > ```
> > # Section: Kernel build options
> > # Summary: Target arch
> > arch=x86_64
> > # Summary: Kernel image name
> > kernel_img_name=bzImage
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You will also need to reorganize all kworkflow.config section a logical way. After that, instead of creating arrays, you can parser the file.
> > What do you think?
> 
> I think it is a good idea, but I have a few remarks:
> 
> * Should we impose another syntax to a configuration file which could be as as simple as possible?
> * Wouldn't it make `kworkflow.config` less user friendly?
> * Your suggestion would imply a bigger feature and demand a bit of time, which I think is not our priority right now;
> * The test is there to help developers remember to update the function whenever they update `kworkflow_template.config` (which shouldn't occur too often, as well)
> 
> What do you think?

I agree, let's not waste time on this feature. Please, tackle the other problems and make a new version of this PR.

File: tests/kw_config_loader_test.sh (hunk 18 above)
Comment (joaoseckler) [reply tests/kw_config_loader_test.sh:151]: Well, the task here is to get what is between parenthesis in the current line. I think it needs a bit of regex anyway...

**Comment by rodrigosiqueira** [comment]
Merged! Thanks
